File: scala/05-ref-path-selection/src/main/scala/org/elu/scala/akka/Counter.scala
package org.elu.scala.akka

import akka.actor.Actor

class Counter extends Actor {
  import org.elu.scala.akka.Counter._

  var count = 0

  override def receive: Receive = {
    case Inc(x) => count += x
    case Dec(x) => count -= x
  }
}

object Counter {
  final case class Inc(num: Int)
  final case class Dec(num: Int)
}

File: kotlin/05-ref-path-selection/src/main/kotlin/org/elu/kotlin/akka/Counter.kt
package org.elu.kotlin.akka

import akka.actor.AbstractActor
import akka.japi.pf.ReceiveBuilder

sealed class Msg
data class Inc(val num: Int) : Msg()
data class Dec(val num: Int) : Msg()

class Counter : AbstractActor() {
  private var count = 0
  override fun createReceive(): Receive =
      ReceiveBuilder()
          .match(Msg::class.java, this::onMessage)
          .build()

  private fun onMessage(msg: Msg) =
      when(msg) {
        is Inc -> count += msg.num
        is Dec -> count -= msg.num
      }
}
